src/lib/common/sol-bus.c

```diff
@@ -397,50 +531,59 @@ _getall_properties(sd_bus_message *reply, void *userdata,
     return _message_map_all_properties(reply, t, ret_error);
 }
 
-int
-sol_bus_map_cached_properties(sd_bus *bus,
-    const char *dest, const char *path, const char *iface,
+SOL_API int
+sol_bus_map_cached_properties(struct sol_bus_client *client,
+    const char *path, const char *iface,
     const struct sol_bus_properties property_table[],
-    void (*changed)(void *data, uint64_t mask),
+    void (*changed)(void *data, const char *path, uint64_t mask),
     const void *data)
 {
     sd_bus_message *m = NULL;
     const struct sol_bus_properties *iter_desc;
     struct property_table *t;
-    char matchstr[4096];
+    char matchstr[512];
     int r;
 
+    SOL_NULL_CHECK(client, -EINVAL);
+
     /* Make sure uint64_t is sufficient to notify state changes - we only
      * support at most 64 properties */
     for (iter_desc = property_table; iter_desc->member != NULL;)
         iter_desc++;
 
-    SOL_INT_CHECK(iter_desc - property_table, >= (int)sizeof(uint64_t), -ENOBUFS);
-
-    r = snprintf(matchstr, sizeof(matchstr),
-        "type='signal',"
-        "sender='%s',"
-        "path='%s',"
-        "interface='org.freedesktop.DBus.Properties',"
-        "member='PropertiesChanged',"
-        "arg0='%s'",
-        dest, path, iface);
-    SOL_INT_CHECK(r, >= (int)sizeof(matchstr), -ENOBUFS);
+    SOL_INT_CHECK(iter_desc - property_table, >= (int)sizeof(uint64_t) * CHAR_BIT, -ENOBUFS);
 
     t = calloc(1, sizeof(*t));
     SOL_NULL_CHECK(t, -ENOMEM);
+
+    t->iface = strdup(iface);
+    SOL_NULL_CHECK_GOTO(t->iface, fail);
+
+    t->path = strdup(path);
+    SOL_NULL_CHECK_GOTO(t->path, fail);
+
     t->properties = property_table;
     t->data = data;
     t->changed = changed;
 
-    r = sol_ptr_vector_append(&_ctx.property_tables, t);
-    SOL_INT_CHECK_GOTO(r, < 0, fail_append);
+    if (!client->properties_changed) {
+        r = snprintf(matchstr, sizeof(matchstr),
+            "type='signal',"
+            "sender='%s',"
+            "interface='org.freedesktop.DBus.Properties',"
+            "member='PropertiesChanged'",
+            client->service);
+        SOL_INT_CHECK_GOTO(r, < 0, fail);
```

Comment:
As I explained in earlier versions of this series, I have a guarantee that `client->service` is smaller than 255 bytes, so the overflow case is not possible.